From 8082052636540808fef65677eaa129e3d32361c7 Mon Sep 17 00:00:00 2001
From: Patrick Lerda <patrick9876@free.fr>
Date: Mon, 10 Nov 2025 12:56:26 +0100
Subject: [PATCH] r600: fix rv770 read scratch compatibility

The flag mega_fetch should be set on rv770 for a
read scratch operation (as written in the r700
documentation p357). Without this flag, read scratch
does not work and a gpu hang could be triggered.

Here are the tests fixed:
shaders/glsl-predication-on-large-array: fail pass
spec/glsl-1.10/execution/temp-array-indexing/glsl-fs-giant-temp-array: fail pass
spec/glsl-1.10/execution/temp-array-indexing/glsl-vs-giant-temp-array: fail pass
spec/glsl-1.30/execution/fs-large-local-array: fail pass
spec/glsl-1.30/execution/fs-large-local-array-vec2: fail pass
spec/glsl-1.30/execution/fs-large-local-array-vec3: fail pass
spec/glsl-1.30/execution/fs-large-local-array-vec4: fail pass
spec/glsl-1.30/execution/fs-multiple-large-local-arrays: fail pass

Fixes: 9c48a139b08f ("r600g: Support emitting scratch ops")
Signed-off-by: Patrick Lerda <patrick9876@free.fr>
Part-of: <https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/38353>
(cherry picked from commit f8de09a811fc1f5b8733e1a36bafd55c778139be)
---
 .pick_status.json                                 | 2 +-
 src/gallium/drivers/r600/ci/r600-rv770-flakes.txt | 1 -
 src/gallium/drivers/r600/ci/r600-rv770-skips.txt  | 1 -
 src/gallium/drivers/r600/r700_asm.c               | 4 +++-
 4 files changed, 4 insertions(+), 4 deletions(-)

diff --git a/.pick_status.json b/.pick_status.json
index 53465bd387e48..d8f292a8d866a 100644
--- a/.pick_status.json
+++ b/.pick_status.json
@@ -124,7 +124,7 @@
         "description": "r600: fix rv770 read scratch compatibility",
         "nominated": true,
         "nomination_type": 2,
-        "resolution": 0,
+        "resolution": 1,
         "main_sha": null,
         "because_sha": "9c48a139b08f223f4ac6e218d19b356bf4a41463",
         "notes": null
diff --git a/src/gallium/drivers/r600/ci/r600-rv770-flakes.txt b/src/gallium/drivers/r600/ci/r600-rv770-flakes.txt
index 5a891237e13ec..57403ae3f2b05 100644
--- a/src/gallium/drivers/r600/ci/r600-rv770-flakes.txt
+++ b/src/gallium/drivers/r600/ci/r600-rv770-flakes.txt
@@ -1,4 +1,3 @@
 spec@!opengl 3.2@gl-3.2-adj-prims pv-last
 spec@!opengl 3.2@layered-rendering@gl-layer-render-clipped
 spec@arb_fragment_layer_viewport@viewport-gs-write-simple
-spec@glsl-1.30@execution@fs-large-local-array-vec4
diff --git a/src/gallium/drivers/r600/ci/r600-rv770-skips.txt b/src/gallium/drivers/r600/ci/r600-rv770-skips.txt
index 8ec2d1bc478dd..5b1e15eff7718 100644
--- a/src/gallium/drivers/r600/ci/r600-rv770-skips.txt
+++ b/src/gallium/drivers/r600/ci/r600-rv770-skips.txt
@@ -7,7 +7,6 @@ texelfetch@gs
 
 # GPU hangs now
 variable-indexing@gs-in
-fs-multiple-large-local-arrays
 
 # intermittent GL_OOMs, let's make sure that doesn't impact other tests.
 spec@!opengl 1.2@tex3d-maxsize
diff --git a/src/gallium/drivers/r600/r700_asm.c b/src/gallium/drivers/r600/r700_asm.c
index 0b1b9f00a0b7b..11e4538b2bac9 100644
--- a/src/gallium/drivers/r600/r700_asm.c
+++ b/src/gallium/drivers/r600/r700_asm.c
@@ -139,7 +139,9 @@ int r700_bytecode_fetch_mem_build(struct r600_bytecode *bc, struct r600_bytecode
 
 	bc->bytecode[id++] = S_SQ_MEM_RD_WORD2_ARRAY_BASE(mem->array_base) |
 		S_SQ_MEM_RD_WORD2_ENDIAN_SWAP(0) |
-		S_SQ_MEM_RD_WORD2_ARRAY_SIZE(mem->array_size);
+		S_SQ_MEM_RD_WORD2_ARRAY_SIZE(mem->array_size) |
+		(bc->gfx_level == R700 && opcode == 0 ?
+		 S_SQ_VTX_WORD2_MEGA_FETCH(1) : 0);
 
 
 	bc->bytecode[id++] = 0; /* MEM ops are 4 word aligned */
-- 
GitLab

